File: frontend/src/services/submissions.ts
import { api } from './api'

/**
 * フォーム送信関連API関数群
 * バックエンドの /submissions エンドポイントと連携
 */

/**
 * 単一フォーム送信リクエスト型
 */
export interface SingleSubmissionRequest {
  form_id: number
  template_id: number
  template_data: Record<string, any>
  take_screenshot?: boolean
  dry_run?: boolean
}

/**
 * 単一フォーム送信レスポンス型
 */
export interface SingleSubmissionResponse {
  message: string
  form_id: number
  template_id: number
  dry_run: boolean
  task_id: string
  status: string
}

/**
 * 送信履歴レスポンス型
 */
export interface SubmissionResponse {
  id: number
  company_id: number
  form_id?: number
  template_id: number
  status: 'pending' | 'success' | 'failed' | 'captcha_required'
  submitted_data: Record<string, any>
  response?: string
  error_message?: string
  submitted_at: string
  screenshot_url?: string
  created_at: string
  updated_at: string
  company_name?: string
  template_name?: string
}

/**
 * 送信履歴一覧レスポンス型
 */
export interface SubmissionListResponse {
  items: SubmissionResponse[]
  total: number
  page: number
  per_page: number
  pages: number
}

/**
 * バッチ送信リクエスト型
 */
export interface BatchSubmissionRequest {
  company_ids: number[]
  template_id: number
  interval_seconds?: number
  test_mode?: boolean
}

/**
 * バッチ送信レスポンス型
 */
export interface BatchSubmissionResponse {
  message: string
  company_count: number
  template_id: number
  interval_seconds: number
  test_mode: boolean
  task_id: string
  status: string
}

/**
 * タスクステータスレスポンス型
 */
export interface TaskStatusResponse {
  task_id: string
  status: 'PENDING' | 'PROGRESS' | 'SUCCESS' | 'FAILURE' | 'REVOKED'
  result?: any
  info?: any
  traceback?: string
}

/**
 * 送信統計レスポンス型
 */
export interface SubmissionStatsResponse {
  period_days: number
  total_submissions: number
  recent_submissions: number
  success_rate: number
  status_breakdown: Record<string, number>
  period: {
    start_date: string
    end_date: string
  }
}

/**
 * 単一フォーム送信を実行
 */
export const submitSingleForm = async (
  request: SingleSubmissionRequest
): Promise<SingleSubmissionResponse> => {
  return await api.post<SingleSubmissionResponse>('/submissions/single', request)
}

/**
 * バッチ送信を実行
 */
export const submitBatchForms = async (
  request: BatchSubmissionRequest
): Promise<BatchSubmissionResponse> => {
  return await api.post<BatchSubmissionResponse>('/submissions/batch', request)
}

/**
 * 送信履歴一覧を取得
 */
export const getSubmissions = async (params?: {
  page?: number
  per_page?: number
  company_id?: number
  status?: string
  start_date?: string
  end_date?: string
}): Promise<SubmissionListResponse> => {
  return await api.get<SubmissionListResponse>('/submissions/', { params })
}

/**
 * 送信履歴詳細を取得
 */
export const getSubmission = async (submissionId: number): Promise<SubmissionResponse> => {
  return await api.get<SubmissionResponse>(`/submissions/${submissionId}`)
}

/**
 * 企業の最新送信履歴を取得
 */
export const getLatestSubmissionByCompany = async (companyId: number): Promise<SubmissionResponse | null> => {
  return await api.get<SubmissionResponse | null>(`/submissions/company/${companyId}/latest`)
}

/**
 * 送信統計を取得
 */
export const getSubmissionStats = async (days: number = 30): Promise<SubmissionStatsResponse> => {
  return await api.get<SubmissionStatsResponse>('/submissions/stats', {
    params: { days }
  })
}

/**
 * 送信履歴を削除
 */
export const deleteSubmission = async (submissionId: number): Promise<void> => {
  await api.delete(`/submissions/${submissionId}`)
}

/**
 * Celeryタスクのステータスを取得
 */
export const getTaskStatus = async (taskId: string): Promise<TaskStatusResponse> => {
  return await api.get<TaskStatusResponse>(`/tasks/${taskId}/status`)
}

/**
 * フォーム送信とタスク監視を統合した関数
 */
export const submitFormAndWaitForResult = async (
  request: SingleSubmissionRequest,
  pollInterval: number = 2000,
  maxAttempts: number = 30
): Promise<{
  success: boolean
  result?: any
  error?: string
  task_id: string
}> => {
  console.group('🚀 フォーム送信プロセス開始')
  console.log(`📝 フォームID: ${request.form_id}`)
  console.log(`📋 テンプレートID: ${request.template_id}`)
  console.log(`🏃 ドライラン: ${request.dry_run ? 'はい' : 'いいえ'}`)
  console.log(`📸 スクリーンショット: ${request.take_screenshot ? 'はい' : 'いいえ'}`)
  console.log(`⏱️ ポーリング間隔: ${pollInterval}ms`)
  console.log(`🎯 最大試行回数: ${maxAttempts}`)
  console.log(`⏰ 開始時刻: ${new Date().toISOString()}`)
  console.groupEnd()

  try {
    // フォーム送信を開始
    const submissionResponse = await submitSingleForm(request)
    
    console.group('📤 フォーム送信レスポンス')
    console.log('🎯 ステータス:', submissionResponse.status)
    console.log('🆔 タスクID:', submissionResponse.task_id)
    console.log('💬 メッセージ:', submissionResponse.message)
    console.groupEnd()

    // タスクの完了を待つ
    for (let attempt = 0; attempt < maxAttempts; attempt++) {
      await new Promise(resolve => setTimeout(resolve, pollInterval))
      
      try {
        const taskStatus = await getTaskStatus(submissionResponse.task_id)
        
        const progress = Math.round(((attempt + 1) / maxAttempts) * 100)
        const remainingTime = Math.round((maxAttempts - attempt - 1) * pollInterval / 1000)
        
        console.group('🔄 タスクステータス確認')
        console.log(`📊 進行状況: ${progress}% (${attempt + 1}/${maxAttempts})`)
        console.log(`⏱️ 残り時間: 約${remainingTime}秒`)
        console.log(`🎯 タスクステータス: ${taskStatus.status}`)
        console.log(`🆔 タスクID: ${submissionResponse.task_id}`)
        console.log(`⏰ タイムスタンプ: ${new Date().toISOString()}`)
        console.groupEnd()
        
        if (taskStatus.status === 'SUCCESS') {
          console.group('✅ フォーム送信完了')
          console.log(`⏱️ 所要時間: ${Math.round((attempt + 1) * pollInterval / 1000)}秒`)
          console.log(`🎯 試行回数: ${attempt + 1}/${maxAttempts}`)
          console.log(`⏰ 完了時刻: ${new Date().toISOString()}`)
          console.log('📋 結果:', taskStatus.result)
          console.groupEnd()
          
          return {
            success: true,
            result: taskStatus.result,
            task_id: submissionResponse.task_id
          }
        } else if (taskStatus.status === 'FAILURE') {
          console.group('❌ フォーム送信失敗')
          console.log(`⏱️ 経過時間: ${Math.round((attempt + 1) * pollInterval / 1000)}秒`)
          console.log(`🎯 試行回数: ${attempt + 1}/${maxAttempts}`)
          console.log(`⏰ 失敗時刻: ${new Date().toISOString()}`)
          console.log('❗ エラー情報:', taskStatus.result)
          if (taskStatus.traceback) {
            console.log('🔍 トレースバック:', taskStatus.traceback)
          }
          console.groupEnd()
          
          return {
            success: false,
            error: taskStatus.result?.error || 'タスクが失敗しました',
            task_id: submissionResponse.task_id
          }
        }
        // PENDING, PROGRESS の場合は続行
      } catch (error) {
        // タスクステータス取得失敗は続行
        console.group('⚠️ タスクステータス取得エラー')
        console.log(`📊 進行状況: ${Math.round(((attempt + 1) / maxAttempts) * 100)}%`)
        console.log(`🆔 タスクID: ${submissionResponse.task_id}`)
        console.log(`❗ エラー詳細: ${error instanceof Error ? error.message : String(error)}`)
        console.log(`⏰ タイムスタンプ: ${new Date().toISOString()}`)
        console.groupEnd()
      }
    }
    
    const totalTimeSeconds = Math.round((maxAttempts * pollInterval) / 1000)
    
    console.group('⏰ フォーム送信タイムアウト')
    console.log(`⌛ 経過時間: ${totalTimeSeconds}秒`)
    console.log(`🎯 試行回数: ${maxAttempts}回`)
    console.log(`🆔 タスクID: ${submissionResponse.task_id}`)
    console.log(`⏰ タイムアウト時刻: ${new Date().toISOString()}`)
    console.log('💡 対処方法:')
    console.log('  • 「再試行」ボタンをクリック')
    console.log('  • 時間をおいてから再実行')
    console.log('  • タスクキューの状態確認')
    console.groupEnd()
    
    return {
      success: false,
      error: `フォーム送信がタイムアウトしました（${totalTimeSeconds}秒経過）。\n\n考えられる原因：\n• サーバーの負荷が高い\n• フォーム送信処理に時間がかかっている\n• ネットワーク接続の問題\n\nタスクID: ${submissionResponse.task_id}`,
      task_id: submissionResponse.task_id
    }

  } catch (error) {
    console.group('❌ フォーム送信開始エラー')
    console.log(`⏰ エラー発生時刻: ${new Date().toISOString()}`)
    console.log(`❗ エラー詳細: ${error instanceof Error ? error.message : String(error)}`)
    console.groupEnd()
    
    throw error
  }
}

/**
 * ドライラン用の簡単な関数
 */
export const dryRunFormSubmission = async (
  formId: number,
  templateId: number,
  templateData: Record<string, any>
): Promise<{
  success: boolean
  result?: any
  error?: string
}> => {
  try {
    const result = await submitFormAndWaitForResult({
      form_id: formId,
      template_id: templateId,
      template_data: templateData,
      take_screenshot: true,
      dry_run: true
    })
    
    return result
  } catch (error) {
    return {
      success: false,
      error: error instanceof Error ? error.message : String(error)
    }
  }
}

/**
 * 実際のフォーム送信用の関数
 */
export const actualFormSubmission = async (
  formId: number,
  templateId: number,
  templateData: Record<string, any>
): Promise<{
  success: boolean
  result?: any
  error?: string
}> => {
  try {
    const result = await submitFormAndWaitForResult({
      form_id: formId,
      template_id: templateId,
      template_data: templateData,
      take_screenshot: true,
      dry_run: false
    })
    
    return result
  } catch (error) {
    return {
      success: false,
      error: error instanceof Error ? error.message : String(error)
    }
  }
}

File: backend/app/api/submissions.py
from typing import Any, List, Optional
from datetime import datetime, timedelta

from fastapi import APIRouter, Depends, HTTPException, Query, BackgroundTasks
from sqlalchemy.orm import Session
from sqlalchemy import and_, desc

from app.api import deps
from app.core.database import get_db
from app.models.submission import Submission
from app.models.company import Company
from app.models.template import Template
from app.models.form import Form
from app.models.user import User
from app.schemas.submission import (
    SubmissionResponse,
    SubmissionListResponse,
    SubmissionBatchCreate,
    SubmissionRequest,
    SubmissionStatus
)
from app.services.form_submitter import form_submitter
from app.tasks.batch_tasks import batch_submission
from app.tasks.form_tasks import submit_form_task

router = APIRouter()


# 古いバッチ送信実装は削除（Celeryタスクに移行）


@router.get("/", response_model=SubmissionListResponse)
def read_submissions(
    *,
    db: Session = Depends(get_db),
    current_user: User = Depends(deps.get_current_active_user),
    page: int = Query(1, ge=1, description="ページ番号"),
    per_page: int = Query(20, ge=1, le=100, description="1ページあたりの件数"),
    company_id: Optional[int] = Query(None, description="企業IDでフィルタ"),
    status: Optional[SubmissionStatus] = Query(None, description="ステータスでフィルタ"),
    start_date: Optional[datetime] = Query(None, description="開始日時"),
    end_date: Optional[datetime] = Query(None, description="終了日時"),
) -> Any:
    """送信履歴一覧を取得"""
    
    query = db.query(Submission)
    
    # フィルタリング
    filters = []
    
    if company_id:
        filters.append(Submission.company_id == company_id)
    
    if status:
        filters.append(Submission.status == status)
    
    if start_date:
        filters.append(Submission.submitted_at >= start_date)
    
    if end_date:
        filters.append(Submission.submitted_at <= end_date)
    
    if filters:
        query = query.filter(and_(*filters))
    
    # 総件数を取得
    total = query.count()
    
    # ページネーション
    offset = (page - 1) * per_page
    submissions = (
        query.order_by(desc(Submission.submitted_at))
        .offset(offset)
        .limit(per_page)
        .all()
    )
    
    # ページ数を計算
    pages = (total + per_page - 1) // per_page
    
    return SubmissionListResponse(
        items=submissions,
        total=total,
        page=page,
        per_page=per_page,
        pages=pages
    )


@router.get("/{submission_id}", response_model=SubmissionResponse)
def read_submission(
    *,
    db: Session = Depends(get_db),
    submission_id: int,
    current_user: User = Depends(deps.get_current_active_user),
) -> Any:
    """送信履歴詳細を取得"""
    submission = db.query(Submission).filter(Submission.id == submission_id).first()
    if not submission:
        raise HTTPException(status_code=404, detail="Submission not found")
    
    return submission


@router.post("/batch", response_model=dict)
async def create_batch_submission(
    *,
    db: Session = Depends(get_db),
    request: SubmissionBatchCreate,
    background_tasks: BackgroundTasks,
    current_user: User = Depends(deps.get_current_active_user),
) -> Any:
    """バッチ送信を実行"""
    
    # 企業IDの妥当性確認
    companies = db.query(Company).filter(Company.id.in_(request.company_ids)).all()
    if len(companies) != len(request.company_ids):
        missing_ids = set(request.company_ids) - {c.id for c in companies}
        raise HTTPException(
            status_code=400,
            detail=f"Companies not found: {list(missing_ids)}"
        )
    
    # テンプレートの存在確認
    template = db.query(Template).filter(Template.id == request.template_id).first()
    if not template:
        raise HTTPException(status_code=404, detail="Template not found")
    
    # Celeryタスクでバッチ送信を実行
    task_result = batch_submission.apply_async(
        args=[
            request.company_ids,
            request.template_id,
            request.interval_seconds,
            request.test_mode,
            True  # take_screenshot
        ]
    )
    
    return {
        "message": "Batch submission started",
        "company_count": len(request.company_ids),
        "template_id": request.template_id,
        "interval_seconds": request.interval_seconds,
        "test_mode": request.test_mode,
        "task_id": task_result.task_id,
        "status": "processing"
    }


@router.post("/single", response_model=dict)
async def create_single_submission(
    *,
    db: Session = Depends(get_db),
    request: SubmissionRequest,
    background_tasks: BackgroundTasks,
    current_user: User = Depends(deps.get_current_active_user),
) -> Any:
    """単一フォーム送信を実行"""
    
    # フォームの存在確認
    form = db.query(Form).filter(Form.id == request.form_id).first()
    if not form:
        raise HTTPException(status_code=404, detail="Form not found")
    
    # テンプレートの存在確認
    template = db.query(Template).filter(Template.id == request.template_id).first()
    if not template:
        raise HTTPException(status_code=404, detail="Template not found")
    
    # Celeryタスクでフォーム送信を実行
    task_result = submit_form_task.apply_async(
        args=[
            request.form_id,
            request.template_id,
            request.template_data,
            request.take_screenshot,
            request.dry_run
        ]
    )
    
    return {
        "message": "Form submission started" if not request.dry_run else "Dry run started",
        "form_id": request.form_id,
        "template_id": request.template_id,
        "dry_run": request.dry_run,
        "task_id": task_result.task_id,
        "status": "processing"
    }


@router.get("/company/{company_id}/latest", response_model=Optional[SubmissionResponse])
def read_latest_submission_by_company(
    *,
    db: Session = Depends(get_db),
    company_id: int,
    current_user: User = Depends(deps.get_current_active_user),
) -> Any:
    """企業の最新送信履歴を取得"""
    submission = (
        db.query(Submission)
        .filter(Submission.company_id == company_id)
        .order_by(desc(Submission.submitted_at))
        .first()
    )
    
    return submission


@router.get("/stats", response_model=dict)
def read_submission_stats(
    *,
    db: Session = Depends(get_db),
    current_user: User = Depends(deps.get_current_active_user),
    days: int = Query(30, ge=1, le=365, description="過去の日数"),
) -> Any:
    """送信統計を取得"""
    
    # 期間を計算
    end_date = datetime.now()
    start_date = end_date - timedelta(days=days)
    
    # 基本統計
    total_submissions = db.query(Submission).count()
    recent_submissions = (
        db.query(Submission)
        .filter(Submission.submitted_at >= start_date)
        .count()
    )
    
    # ステータス別統計
    status_stats = {}
    for status in SubmissionStatus:
        count = (
            db.query(Submission)
            .filter(
                and_(
                    Submission.status == status,
                    Submission.submitted_at >= start_date
                )
            )
            .count()
        )
        status_stats[status.value] = count
    
    # 成功率
    success_count = status_stats.get(SubmissionStatus.SUCCESS.value, 0)
    success_rate = (success_count / recent_submissions * 100) if recent_submissions > 0 else 0
    
    return {
        "period_days": days,
        "total_submissions": total_submissions,
        "recent_submissions": recent_submissions,
        "success_rate": round(success_rate, 2),
        "status_breakdown": status_stats,
        "period": {
            "start_date": start_date.isoformat(),
            "end_date": end_date.isoformat()
        }
    }


@router.delete("/{submission_id}")
def delete_submission(
    *,
    db: Session = Depends(get_db),
    submission_id: int,
    current_user: User = Depends(deps.get_current_active_user),
) -> Any:
    """送信履歴を削除"""
    submission = db.query(Submission).filter(Submission.id == submission_id).first()
    if not submission:
        raise HTTPException(status_code=404, detail="Submission not found")
    
    db.delete(submission)
    db.commit()
    
    return {"message": "Submission deleted successfully"}